### PR TITLE
test: vaults fuzzing

### PR DESCRIPTION
**Pull request by sergeyWh1te**
Test shareRate with fuzzing

### CODE REVIEW

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol [hunk 1]
```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "../../../../contracts/0.8.9/EIP712StETH.sol";
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]:
```suggestion
import "contracts/0.8.9/EIP712StETH.sol";
```

It's possible to use shortcut

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+import "forge-std/Test.sol";
+
+import {CommonBase} from "forge-std/Base.sol";
+import {LidoLocator} from "../../../../contracts/0.8.9/LidoLocator.sol";
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]:
```suggestion
import {LidoLocator} from "contracts/0.8.9/LidoLocator.sol";
```

Probably reorder internal imports to single place (top / bottom) for clarity

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import {StdCheats} from "forge-std/StdCheats.sol";
+import {StdUtils} from "forge-std/StdUtils.sol";
+import {Vm} from "forge-std/Vm.sol";
+import {console2} from "forge-std/console2.sol";
+
+interface ILido {
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]: Can't we use an already defined interface from `contracts/0.8.25/interfaces/..`?

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    address private rootAccount;
+    address private userAccount;
+
+    address public kernelBase;
+    address public aclBase;
+    address public evmScriptRegistryFactory;
+    address public daoFactoryAdr;
+
+    address public accounting = makeAddr("dummy-locator:accounting");
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]: Do we need to separate only accounting? Maybe it's better to expose the locator to tests and then just call the required contracts from the locator? (see below)

File: test/0.8.25/vaults/contracts/ShareRate.t.sol [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            userAccount
+        );
+
+        vm.startPrank(userAccount);
+        lidoContract.setMaxExternalRatioBP(_maxExternalRatioBP);
+        lidoContract.setStakingLimit(_maxStakeLimit, _stakeLimitIncreasePerBlock);
+        lidoContract.resume();
+        vm.stopPrank();
+
+        shareRateHandler = new ShareRateHandler(lidoContract, accounting, userAccount, _maxAmountOfShares);
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/ShareRate.t.sol]: Here it may be 
```suggestion
        shareRateHandler = new ShareRateHandler(lidoContract, locator.accounting(), userAccount, _maxAmountOfShares);
```

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+    function prepareLidoContract(uint256 _startBalance, address _rootAccount, address _userAccount) public {
+        rootAccount = _rootAccount;
+        userAccount = _userAccount;
+
+        vm.startPrank(rootAccount);
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]: Maybe separate those "pranks" to a separate methods for better readability? like setting up Kernet, setup dao, setup lido, etc...

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+        vm.recordLogs();
+        daoFactory.newDAO(rootAccount);
+        Vm.Log[] memory logs = vm.getRecordedLogs();
+        (address daoAddress) = abi.decode(logs[logs.length - 1].data, (address));
+
+        vm.startPrank(rootAccount);
+            IKernel dao = IKernel(address(daoAddress));
+            IACL acl = IACL(address(dao.acl()));
+            acl.createPermission(rootAccount, daoAddress, keccak256("APP_MANAGER_ROLE"), rootAccount);
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]: expose helper methods in this contract to add/remove permission?

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol [hunk 8]
```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: MIT
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]:
```suggestion
// SPDX-License-Identifier: UNLICENSED
// for testing purposes only

```

File: test/0.8.25/vaults/contracts/ShareRate.t.sol [hunk 9]
```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: MIT
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/ShareRate.t.sol]:
```suggestion
// SPDX-License-Identifier: UNLICENSED
// for testing purposes only

```

File: test/0.8.25/vaults/contracts/ShareRate.t.sol [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+import "../../../../contracts/0.8.9/EIP712StETH.sol";
+import "forge-std/Test.sol";
+
+import {CommonBase} from "forge-std/Base.sol";
+import {LidoLocator} from "../../../../contracts/0.8.9/LidoLocator.sol";
+import {StdCheats} from "forge-std/StdCheats.sol";
+import {StdUtils} from "forge-std/StdUtils.sol";
+import {Vm} from "forge-std/Vm.sol";
+import {console2} from "forge-std/console2.sol";
+import {Protocol__Deployment, ILido} from "./Protocol__Deployment.t.sol";
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/ShareRate.t.sol]: Maybe better naming is Protocol__Context so as to put methods related to the protocol interactions later, e.g., adding / removing permissions, impersonating agents, etc...

File: test/0.8.25/vaults/contracts/ShareRate.t.sol [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        lidoContract = _lido;
+        accounting = _accounting;
+        userAccount = _userAccount;
+        maxAmountOfShares = _maxAmountOfShares;
+    }
+
+    function mintExternalShares(address _recipient, uint256 _amountOfShares) external {
+        _amountOfShares = bound(_amountOfShares, 1, maxAmountOfShares);
+
+        vm.startPrank(userAccount);
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/ShareRate.t.sol]:
```suggestion
        vm.prank(userAccount);
```

In that case no need to call `stopPrank` and readability increases, imo

File: test/0.8.25/vaults/contracts/ShareRate.t.sol [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        lidoContract.resumeStaking();
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/ShareRate.t.sol]: This may be done in setup action

File: test/0.8.25/vaults/contracts/ShareRate.t.sol [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+        vm.startPrank(accounting);
+        lidoContract.mintExternalShares(_recipient, _amountOfShares);
+        vm.stopPrank();
+    }
+
+    function burnExternalShares(uint256 _amountOfShares) external {
+        _amountOfShares = bound(_amountOfShares, 1, maxAmountOfShares);
+        vm.startPrank(userAccount);
+        lidoContract.resumeStaking();
```
Comment (tamtamchik) [inline test/0.8.25/vaults/contracts/ShareRate.t.sol]: Same here, move to setup

**Review by tamtamchik — COMMENTED** [review]
Great start! 

Some suggestions and comments left.

Also, looks like it need to me merged with vaults branch once again, because some tests are failing.

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol (hunk 1 above)
Comment (sergeyWh1te) [reply test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]: refactored

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol (hunk 2 above)
Comment (sergeyWh1te) [reply test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]: refactored

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol (hunk 3 above)
Comment (sergeyWh1te) [reply test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]: Unfortunately, no
contracts/0.8.25/interfaces/ILido does not contain: 
```
setMaxExternalRatioBP
initialize
resumeStaking
resume
setStakingLimit
```

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol (hunk 4 above)
Comment (sergeyWh1te) [reply test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]: refactored

File: test/0.8.25/vaults/contracts/ShareRate.t.sol (hunk 5 above)
Comment (sergeyWh1te) [reply test/0.8.25/vaults/contracts/ShareRate.t.sol]: refactored

File: test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol (hunk 8 above)
Comment (sergeyWh1te) [reply test/0.8.25/vaults/contracts/Protocol__Deployment.t.sol]: refactored

File: test/0.8.25/vaults/contracts/ShareRate.t.sol (hunk 9 above)
Comment (sergeyWh1te) [reply test/0.8.25/vaults/contracts/ShareRate.t.sol]: refactored

File: test/0.8.25/Accounting.t.sol [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+        uint256 _clBalanceGwei,
+        uint256 _withdrawalVaultBalanceWei,
+        uint256 _totalDepositedValidators,
+        uint256 _totalExitedValidators
+    ) external;
+}
+
+// 0.002792 * 10^18
+// 0.0073 * 10^18
+uint256 constant maxYiedPerOperatorWei = 2_792_000_000_000_000; // which % of slashing could be?
```
Comment (tamtamchik) [inline test/0.8.25/Accounting.t.sol]:
```suggestion
uint256 constant maxYieldPerOperatorWei = 2_792_000_000_000_000; // which % of slashing could be?
```
?

File: test/0.8.25/Accounting.t.sol [hunk 15]
```diff
@@ -0,0 +1,360 @@
+// SPDX-License-Identifier: UNLICENSED
+// for testing purposes only
+pragma solidity ^0.8.0;
+
+import "foundry/lib/forge-std/src/Vm.sol";
```
Comment (tamtamchik) [inline test/0.8.25/Accounting.t.sol]:
```suggestion
```
Already present below

File: test/0.8.25/Accounting.t.sol [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+import {CommonBase} from "forge-std/Base.sol";
+import {StdCheats} from "forge-std/StdCheats.sol";
+import {StdUtils} from "forge-std/StdUtils.sol";
+import {Vm} from "forge-std/Vm.sol";
+import {console2} from "../../foundry/lib/forge-std/src/console2.sol";
```
Comment (tamtamchik) [inline test/0.8.25/Accounting.t.sol]:
```suggestion
import {console2} from "forge-std/console2.sol";
```

File: test/0.4.24/contracts/SecondOpinionOracle__Mock.sol
Comment (tamtamchik) [inline test/0.4.24/contracts/SecondOpinionOracle__Mock.sol:1]: Better to put it into test/0.8.25/contracts and rename it like SecondOpinionOracle__MockForAccountingFuzzing

File: test/0.4.24/contracts/StakingRouter__MockForLidoAccounting.sol [hunk 17]
```diff
@@ -3,8 +3,11 @@
 
 pragma solidity 0.8.9;
 
+import {StakingRouter} from "contracts/0.8.9/StakingRouter.sol";
+
 contract StakingRouter__MockForLidoAccounting {
```
Comment (tamtamchik) [inline test/0.4.24/contracts/StakingRouter__MockForLidoAccounting.sol]: Noticed that this file is being used in the accounting unit tests. It would be better to clone the file and rename it to StakingRouter__MockForLidoAccountingFuzzing to clarify its purpose. Separating the mocks will simplify mock management and ensure that they do not interfere with other tests.

File: test/0.4.24/contracts/StakingRouter__MockForLidoAccounting.sol [hunk 18]
```diff
@@ -47,4 +57,75 @@ contract StakingRouter__MockForLidoAccounting {
...
+    function getRecipients() public view returns (address[] memory) {
+        return recipients__mocked;
+    }
+
+    function getStakingModule(uint256 _stakingModuleId) public view returns (StakingRouter.StakingModule memory) {
+        if (_stakingModuleId >= 4) {
+            revert("Staking module does not exist");
+        }
+
+        if (_stakingModuleId == 1) {
```
Comment (tamtamchik) [inline test/0.4.24/contracts/StakingRouter__MockForLidoAccounting.sol]: Maybe add a comment that production values are used? And if possible to add a proof link. Otherwise no one will guess why the numbers are like they are here.

File: test/0.8.25/Protocol__Deployment.t.sol [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+            maxItemsPerExtraDataTransaction: 8,
+            maxNodeOperatorsPerExtraDataItem: 24,
+            requestTimestampMargin: 7680,
+            maxPositiveTokenRebase: 750000,
+            initialSlashingAmountPWei: 1000,
+            inactivityPenaltiesAmountPWei: 101,
+            clBalanceOraclesErrorUpperBPLimit: 50
+        });
+
+    function setUpProtocol(uint256 _startBalance, address _rootAccount, address _userAccount) public {
```
Comment (tamtamchik) [inline test/0.8.25/Protocol__Deployment.t.sol:119]: It seems that we can divide this into separate deployments since we now have two tests that might interfere with the setup. Perhaps we don't need a universal deployment at all? Instead just a bunch of deployment actions to call from test's setUp function? This is something we need to consider.

File: test/0.8.25/Protocol__Deployment.t.sol [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+
+        uint96[] memory stakingModuleFees = new uint96[](3);
+        stakingModuleFees[0] = 4876942047684326532;
+        stakingModuleFees[1] = 145875332634464962;
+        stakingModuleFees[2] = 38263043302959438;
+
+        address[] memory recipients = new address[](3);
+        recipients[0] = 0x55032650b14df07b85bF18A3a3eC8E0Af2e028d5;
+        recipients[1] = 0xaE7B191A31f627b4eB1d4DaC64eaB9976995b433;
+        recipients[2] = 0xdA7dE2ECdDfccC6c3AF10108Db212ACBBf9EA83F;
```
Comment (tamtamchik) [inline test/0.8.25/Protocol__Deployment.t.sol:154]: Not sure why this is duplicated both here and in the mock itself. What if the number is changed somewhere and we end up with a discrepancy? Maybe we should find a way to unify it.

File: test/0.8.25/ShareRate.t.sol [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+        lidoContract.mintExternalShares(accountingContract, protocolStartExternalShares);
+    }
+
+    /**
+     * https://book.getfoundry.sh/reference/config/inline-test-config#in-line-invariant-configs
+     * forge-config: default.invariant.runs = 256
+     * forge-config: default.invariant.depth = 256
+     * forge-config: default.invariant.fail-on-revert = true
+     */
+    function invariant_totalShares() public view {
```
Comment (tamtamchik) [inline test/0.8.25/ShareRate.t.sol:118]: 💡 Maybe add an invariant that lido.getExternalShares = startExternalBalance + mintedExternal - burnedExternal? So we'll know it something is odd inside a math for external shares?

File: test/0.8.25/Accounting.t.sol [hunk 22]
```diff
@@ -0,0 +1,360 @@
...
+
+    function handleOracleReport(FuzzValues memory fuzz) external {
+        uint256 _timeElapsed = 86_400;
+        uint256 _timestamp = 1_737_366_566 + _timeElapsed;
+
+        // cheatCode for
+        // if (_report.timestamp >= block.timestamp) revert IncorrectReportTimestamp(_report.timestamp, block.timestamp);
+        vm.warp(_timestamp + 1);
+
+        fuzz._lidoExecutionLayerRewardVaultWei = bound(fuzz._lidoExecutionLayerRewardVaultWei, 0, 1_000) * 1 ether;
```
Comment (tamtamchik) [inline test/0.8.25/Accounting.t.sol:163]: Providing comments on why the bounding appears as it does would be helpful during reviews and test updates.

File: test/0.8.25/Accounting.t.sol [hunk 23]
```diff
@@ -0,0 +1,360 @@
...
+        stakingRouter = IStakingRouter(_stakingRouter);
+    }
+
+    function cutGwei(uint256 value) public returns (uint256) {
+        return (value / 1 gwei) * 1 gwei;
+    }
+
+    function handleOracleReport(FuzzValues memory fuzz) external {
+        uint256 _timeElapsed = 86_400;
+        uint256 _timestamp = 1_737_366_566 + _timeElapsed;
```
Comment (tamtamchik) [inline test/0.8.25/Accounting.t.sol]: Why do we need a specific timestamp? Can block.timestamp be used?

File: test/0.8.25/Accounting.t.sol [hunk 24]
```diff
@@ -0,0 +1,360 @@
...
+    uint256 _lidoExecutionLayerRewardVaultWei;
+}
+
+struct LidoTransfer {
+    address from;
+    address to;
+}
+
+contract AccountingHandler is CommonBase, StdCheats, StdUtils {
+    struct Ghost {
```
Comment (tamtamchik) [inline test/0.8.25/Accounting.t.sol:78]: 💡 Maybe add a boundary holder to track minimum and maximum values used in all the runs, so we can review them after the test and ensure we're not missing any potential values chunks?

File: test/0.4.24/contracts/SecondOpinionOracle__Mock.sol [hunk 25]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: UNLICENSED
+// for testing purposes only
+
+pragma solidity 0.8.9;
+
+contract SecondOpinionOracle__Mock {
```
Comment (tamtamchik) [inline test/0.4.24/contracts/SecondOpinionOracle__Mock.sol]: Not sure that we need that mock. We can harness OracleReportSanityChecker and just return true on _askSecondOpinion.

**Review by tamtamchik — COMMENTED** [review]
No major issues, at least for now, need to validate boundaries and maybe add more invariants.

File: test/0.8.25/Accounting.t.sol (hunk 14 above)
Comment (sergeyWh1te) [reply test/0.8.25/Accounting.t.sol]: fixed

File: test/0.8.25/Accounting.t.sol (hunk 15 above)
Comment (sergeyWh1te) [reply test/0.8.25/Accounting.t.sol]: fixed

File: test/0.8.25/Accounting.t.sol (hunk 16 above)
Comment (sergeyWh1te) [reply test/0.8.25/Accounting.t.sol]: fixed

File: test/0.4.24/contracts/StakingRouter__MockForLidoAccounting.sol (hunk 17 above)
Comment (sergeyWh1te) [reply test/0.4.24/contracts/StakingRouter__MockForLidoAccounting.sol]: fixed

File: test/0.8.25/Accounting.t.sol (hunk 23 above)
Comment (sergeyWh1te) [reply test/0.8.25/Accounting.t.sol]: replaced

File: test/0.8.25/Accounting.t.sol (hunk 24 above)
Comment (sergeyWh1te) [reply test/0.8.25/Accounting.t.sol:78]: ?

File: test/0.4.24/contracts/SecondOpinionOracle__Mock.sol (hunk 25 above)
Comment (sergeyWh1te) [reply test/0.4.24/contracts/SecondOpinionOracle__Mock.sol]: ?

File: test/0.8.25/ShareRate.t.sol (hunk 21 above)
Comment (sergeyWh1te) [reply test/0.8.25/ShareRate.t.sol:118]: added like TODO

File: test/0.8.25/Protocol__Deployment.t.sol (hunk 20 above)
Comment (sergeyWh1te) [reply test/0.8.25/Protocol__Deployment.t.sol:154]: will optimize

File: test/0.8.25/Accounting.t.sol (hunk 24 above)
Comment (tamtamchik) [reply test/0.8.25/Accounting.t.sol:78]: Added in e9a302d4175e9a3217d30594bdd58d834c7452c8. Shows the max <> min range.. Like this:
```
  Boundary Values:
  PreClValidators min: 250000
  PreClValidators max: 100000000000
  ClValidators min: 259381
  ClValidators max: 100000043199
  ClBalanceWei min: 8300705805470639196490834
  ClBalanceWei max: 3200271341074368560000000000000
  DepositedValidators min: 293200
  DepositedValidators max: 100000086397
  ElRewardsVaultBalanceWei min: 0
  ElRewardsVaultBalanceWei max: 977277467356288757093
```

File: test/0.4.24/contracts/SecondOpinionOracle__Mock.sol (hunk 25 above)
Comment (tamtamchik) [reply test/0.4.24/contracts/SecondOpinionOracle__Mock.sol]: Done in e9a302d4175e9a3217d30594bdd58d834c7452c8

**Comment by tamtamchik** [comment]
Development moved to #952
